### ISSUE TITLE
Switch to autotools

### ISSUE THREAD

**Issue by Jehan** [mypaint/libmypaint#32]
To update my Windows cross-build, I need to cross-compile libmypaint, which is now a mandatory dependency.
Unfortunately I was unable to do so. Well I didn't try much yet, but I at least tried to set $CC, and other environment variable (as I saw in the SConstruct), and used my usual cross-build environment with other environment variables useful for autotools as well as CMake crossbuilds. It was not enough.
Since you also provide MyPaint for Windows, do you have cross-compilation instructions somewhere? I did not find any in the repository.

As an alternative, I am wondering if you would consider switching to another build system which has better support for cross-compilation in a standard way (common to all projects using it), for instance the autotools?

**Comment by tumagonx** [comment]
Post your error message :) 
Libmypaint.dll alone  should be easy to made in plain makefile though. Heck, I used to invoke oneliner gcc manually :) What complex is translation string generation and introspection (which maybe need run wine python natively).

**Comment by Jehan** [comment]
I retried with a completely clean repo and tried to debug a little more this time.

At first, I had an issue with a warning which seems rather common when cross-compiling, but is rather harmless: -fPIC is a useless option for building for Windows apparently, but libtool always tries to add it, and I can't find a way to get rid of it (it has been going for years and a [patch to gcc](https://gcc.gnu.org/ml/gcc-patches/2015-08/msg00836.html) recently [got rid](https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff;h=956d69de43315a938beb07d8a379f2c558a87158) of the warning.

> scons: Reading SConscript files ...
> scons: done reading SConscript files.
> scons: Building targets ...
> Creating 'mypaint-config.h'
> x86_64-w64-mingw32-gcc -o brushmodes.os -c -std=c99 -D_POSIX_C_SOURCE=200809L -fopenmp -O3 -Wall -Wstrict-prototypes -Werror -g -fPIC -DHAVE_JSON_C -DHAVE_GETTEXT -I. -I/usr/include/json-c brushmodes.c
> brushmodes.c:1:0: error: -fPIC ignored for target (all code is position independent) [-Werror]
 > /* brushlib - The MyPaint Brush Library
 > ^
> cc1: all warnings being treated as errors
> scons: *** [brushmodes.os] Error 1
> scons: building terminated because of errors.

In the meantime, I temporary disabled -Werror to continue the build.
Next error:

> x86_64-w64-mingw32-gcc -o tests/test-brush-persistence -O3 -fopenmp tests/test-brush-persistence.o -L. -lmypaint-tests -lmypaint -lm -ljson-c
> x86_64-w64-mingw32-gcc: error: libgomp.spec: No such file or directory
> scons: *** [tests/test-brush-persistence] Error 1
> scons: building terminated because of errors.

Some search told me it would be linked to openmp, so I tried enable_openmp=no but then I just got a different error:

> x86_64-w64-mingw32-gcc -o tests/test-brush-persistence -O3 tests/test-brush-persistence.o -L. -lmypaint-tests -lmypaint -lm -ljson-c
> ./libmypaint.a: error adding symbols: Archive has no index; run ranlib to add one
> collect2: error: ld returned 1 exit status
> scons: *** [tests/test-brush-persistence] Error 1

In the end, I just commented out the tests/SConscript in the main SConscript (a "tests/" directory, I assume these are unit tests. And you are right, I managed to build libmypaint. But this could be less random. scons has apparently no concept of host/target nor anything related to cross-compilation which makes it really unreliable in such a context, IMHO.

Well still I found workarounds. I guess this issue could be closed with -Werror removed when a Win32 cross-compilation is detected, and with the tests not built either on cross-compilation detected. Unless you know how to fix these warnings/errors of course.

**Comment by tumagonx** [comment]
which python you use? wine or linux?

**Comment by Jehan** [comment]
linux.

**Comment by Jehan** [comment]
linux.

**Comment by tumagonx** [comment]
That would be unfortunate (see [site-packages]\Scons\tool\gcc.py) in the end you still need native python for introspection (if this eventually enabled)...

**Comment by tumagonx** [comment]
this post may help https://bitbucket.org/scons/scons/wiki/CrossCompilingMingw
but doesn't solve introspection

**Comment by Jehan** [comment]
Actually I was wrong. I only manage to build libmypaint with use_sharedlib=no. And then linking with GIMP fails with a similar error as I had building libmypaint tests:

> libmypaint-gegl.a: error adding symbols: Archive has no index; run ranlib to add one

When I set use_sharedlib=yes, scons build says:

> scons: Building targets ...
> x86_64-w64-mingw32-gcc -o libmypaint-gegl.so -O3 -Wl,--export-dynamic -pthread -shared gegl/mypaint-gegl-surface.os -L/home/jehan/.local/lib -L. -ljson-c -lgegl-0.3 -lgegl-npd-0.3 -lm -lgmodule-2.0 -ljson-glib-1.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lbabl-0.1 -lmypaint
> /usr/lib/gcc/x86_64-w64-mingw32/5.2.0/../../../../x86_64-w64-mingw32/bin/ld: warning: --export-dynamic is not supported for PE+ targets, did you mean --export-all-symbols?
> ./libmypaint.a: error adding symbols: Archive has no index; run ranlib to add one
> collect2: error: ld returned 1 exit status
> scons: *** [libmypaint-gegl.so] Error 1

Why is there even a message about libmypaint.a when I asked for a shared library? And why does the build command tries to output a .so and not a .dll?

As for the introspection, it is not too big a problem right now, since we don't use it in GIMP. I would settle for a working build without introspection for the time being.

**Comment by tumagonx** [comment]
Could you post full log (in pastebin)?
that libmypaint.a (probably should be libmypaint.dll.a) is import lib. Again, note that because of you use linux python your flags seems still all wrong

**Comment by tumagonx** [comment]
BTW, I recall long time ago when I built libmypaint with gegl it crashed mypaint. I think Jonnor know how gegl situation in libmypaint,  can't test it myself right now.

**Comment by achadwick** [comment]
I definitely see autotools in libmypaint's future, post-split (mypaint/mypaint#535, #28, #29). If it's any better with its cross-compilation support, that's just another point in its favour.

**Comment by Jehan** [comment]
@tumagonx > I tested a mypaint build with libmypaint with gegl a few days ago and it didn't crash. Now I have not extensively tested so if the crash reproduction is subtle, I may just not have encountered it. But basic opening MyPaint, drawing a bit and playing around worked fair.

For the full log: http://fpaste.org/305470/31413314/

@achadwick > well it has cross-compilation support by default, yes. All you have to do is follow clean standards accross your Makefile.am, like never call 'gcc' or 'pkg-config' directly, but use proper wrapper functions (like PKG_CHECK_EXISTS()) or variables when needed (like $(CC), $PKG_CONFIG) and so on.
I am usually able to cross-build well-written autotools projects as easily as I am to build them.

I'm happy to know you see autotools in libmypaint's future. :-)

**Comment by tumagonx** [comment]
@achadwick there is two issues with current scons
1. libs = ['intl'] need to be added conditionally for windows if gettext enabled (I don't knoe wih linux)
2. install didn't copy the dll files (mypaint.dll, mypaint_gegl.dll, mypaint_test.dll) probably right now it look for *.so files?

@jehan you test Windows version? thats great! probably final gegl 0.3 fixed it, I'm pretty I used gegl git two years ago. Note: Your log obviously "mingw in unix way" :) the use of .so extension, .os (shared object) and those wrong flags. Scons has a draft called IAPAT to support cross-compile by allow override "PLATFORM" env.  Well, just a draft yet.

Now I'm bit confused how to enable gegl? I remember back then I need to pass USE_GEGL (or somethin) environment to mypaint.exe to enable it now all I can see is gegl.py file with no reference by other. Though I have all files built, (mypaint typelibs, geglgtk3 etc.) I still can't see gegl module loaded.

**Comment by Jehan** [comment]
> you test Windows version?

Sorry to mislead you, but no I don't test MyPaint for Windows, if that's what you mean. I only rarely (like once every X months) hack or test *GIMP* on Windows since we don't have many dedicated Windows developers unfortunately. Maybe I misunderstood when you said "I built libmypaint with gegl it crashed mypaint". Did you mean "it crashed MyPaint on Windows"? Because I haven't tested MyPaint on Windows, on Linux only.
Also I haven't set any environment variable either. I just ran MyPaint with libmypaint built with GEGL. Nothing more. I have no idea what is the difference for MyPaint code and haven't checked to be clear.
Sorry to disappoint you. :P

**Comment by Jehan** [comment]
Ok @achadwick so I have mostly ported to autotools, as planned. Actually most of the works was done during LGM, but I had to do some fixes here and there.
Some code is coming from GIMP, for instance the platform detection code.
For now this is available in the branch `autotools-port`, not merged to master yet.

Now libmypaint uses autotools convention.

(1) So you have to kickstart the build environment with:
$ ./autogen.sh
Note that it also runs the first configure. If you don't want this to happen, export the $NOCONFIGURE environment variable.
This autogen.sh was mostly taken from the GIMP project because it just works. :-)
This script is for the hackers who want to generate the configure script the first time from configure.ac. Users won't have to deal with it and will have the configure generated from the start.

(2) Now run:
$ ./configure
As usual, add any of the common GNU options, like --prefix, --host (for cross-compilation), --enable-shared/static to build whatever you want, --bindir/libdir/includedir/whateverdir for any hardcore user who like to play with one's environment and so on.
Then the libmypaint specific options: --enable-docs/debug/profiling/openmp/gperftools/i18n/introspection/gegl and --with-glib.
You get the full list with `./configure -h`.

(3) Then `make` and `make install`.

(4) After you can run all the unit tests with `make check` (works now with both shared and static libs), update the po files from the source code with `cd po && make update-po`. Common/practical GNU stuff. :-)
List of language is maintained through po/LINGUAS (currently it has all the lang, I left a comment on the top of the files with a command line to get the full list of languages. But you can also disable languages by removing them from the list if needed).
List of files where localizable strings can be found is maintained through po/POTFILES.

Now I am just 100% sure that there are issues here and there. This was not the tougher port, but also not the easier, mostly because there are a lot of broken rules in the scons build, and stuff I am still not sure about. Also the other problem is that some part of the code are harder to test. GIMP does not use mypaint-gegl nor the introspection for instance. MyPaint neither, I think.

But I propose to start with this and fix things as they go.
At least I can confirm that GIMP builds and the mypaint brushes seem to work correctly.
So if you could review the branch `autotools-port` then if all is good, I will merge it, that would be cool. :-)

**Comment by achadwick** [comment]
@Jehan
Will do. Thanks for preparing this!

**Comment by achadwick** [comment]
I've pushed a few changes to capture those docs and make it build on my system.

Down to one error during `autogen.sh`:

    sed: can't read ./po/POTFILES.in: No such file or directory
 
Hasn't it changed to just `po/POTFILES` now?

#### mypaint-config.h is not generated

`make` Fails to build with a missing `mypaint-config.h`:

```$ make
make  all-recursive
make[1]: Entering directory '/home/andrewc/Development/MyPaintProject/libmypaint'
Making all in .
make[2]: Entering directory '/home/andrewc/Development/MyPaintProject/libmypaint'
depbase=`echo brushmodes.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
/bin/bash ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I.    -I/usr/include/json-c  -D_POSIX_C_SOURCE=200809L -g -O2 -MT brushmodes.lo -MD -MP -MF $depbase.Tpo -c -o brushmodes.lo brushmodes.c &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/usr/include/json-c -D_POSIX_C_SOURCE=200809L -g -O2 -MT brushmodes.lo -MD -MP -MF .deps/brushmodes.Tpo -c brushmodes.c  -fPIC -DPIC -o .libs/brushmodes.o
In file included from rng-double.h:4:0,
                 from helpers.h:4,
                 from brushmodes.c:20:
./mypaint-glib-compat.h:4:28: fatal error: mypaint-config.h: No such file or directory
compilation terminated.
Makefile:712: recipe for target 'brushmodes.lo' failed
make[2]: *** [brushmodes.lo] Error 1
make[2]: Leaving directory '/home/andrewc/Development/MyPaintProject/libmypaint'
Makefile:843: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/andrewc/Development/MyPaintProject/libmypaint'
Makefile:580: recipe for target 'all' failed
make: *** [all] Error 2
```

#### cleanup targets not supported or broken

`make maintainer-clean` is failing in `po`:

```
make[1]: *** No rule to make target 'POTFILES.in', needed by 'stamp-it'.  Stop.
make[1]: Leaving directory '/home/andrewc/Development/MyPaintProject/libmypaint/po'
Makefile:843: recipe for target 'maintainer-clean-recursive' failed
make: *** [maintainer-clean-recursive] Error 1
```

The `clean` and `distclean` targets are not supported in `doc/`.

**Comment by Jehan** [comment]
> Hasn't it changed to just po/POTFILES now?

Hmmm... Indeed in GIMP, there is a po/POTFILES.in. So I guess I made an error!

> make Fails to build with a missing mypaint-config.h:

Ok I'll check. I guess my main mistake was that I didn't make a full cleaning of all generated files at some point to test with a clean repo (similar to a just-cloned repo).

> The clean and distclean targets are not supported in doc/.

Which makes sense since the doc building is fully manual (not using standard doc generation rules. I wrote them myself), so autotools does not know exactly what to clean either. I will need to manually write the clean/distclean rules, which I didn't.

I will have a look at these 3 issues later (today if I can).

By the way, the docs are not installed either. They are only built, because this was already the case in your old scons build. So I just mirrored this behavior. Don't you want these installed? For instance in $(datadir)/docs/libmypaint/?

**Comment by Jehan** [comment]
> Hasn't it changed to just po/POTFILES now?

Hmmm... Indeed in GIMP, there is a po/POTFILES.in. So I guess I made an error!

> make Fails to build with a missing mypaint-config.h:

Ok I'll check. I guess my main mistake was that I didn't make a full cleaning of all generated files at some point to test with a clean repo (similar to a just-cloned repo).

> The clean and distclean targets are not supported in doc/.

Which makes sense since the doc building is fully manual (not using standard doc generation rules. I wrote them myself), so autotools does not know exactly what to clean either. I will need to manually write the clean/distclean rules, which I didn't.

I will have a look at these 3 issues later (today if I can).

By the way, the docs are not installed either. They are only built, because this was already the case in your old scons build. So I just mirrored this behavior. Don't you want these installed? For instance in $(datadir)/docs/libmypaint/?

**Comment by Jehan** [comment]
> Hasn't it changed to just po/POTFILES now?

Hmmm... Indeed in GIMP, there is a po/POTFILES.in. So I guess I made an error!

> make Fails to build with a missing mypaint-config.h:

Ok I'll check. I guess my main mistake was that I didn't make a full cleaning of all generated files at some point to test with a clean repo (similar to a just-cloned repo).

> The clean and distclean targets are not supported in doc/.

Which makes sense since the doc building is fully manual (not using standard doc generation rules. I wrote them myself), so autotools does not know exactly what to clean either. I will need to manually write the clean/distclean rules, which I didn't.

I will have a look at these 3 issues later (today if I can).

By the way, the docs are not installed either. They are only built, because this was already the case in your old scons build. So I just mirrored this behavior. Don't you want these installed? For instance in $(datadir)/docs/libmypaint/?

**Comment by Jehan** [comment]
> Hasn't it changed to just po/POTFILES now?

Hmmm... Indeed in GIMP, there is a po/POTFILES.in. So I guess I made an error!

> make Fails to build with a missing mypaint-config.h:

Ok I'll check. I guess my main mistake was that I didn't make a full cleaning of all generated files at some point to test with a clean repo (similar to a just-cloned repo).

> The clean and distclean targets are not supported in doc/.

Which makes sense since the doc building is fully manual (not using standard doc generation rules. I wrote them myself), so autotools does not know exactly what to clean either. I will need to manually write the clean/distclean rules, which I didn't.

I will have a look at these 3 issues later (today if I can).

By the way, the docs are not installed either. They are only built, because this was already the case in your old scons build. So I just mirrored this behavior. Don't you want these installed? For instance in $(datadir)/docs/libmypaint/?

**Comment by Jehan** [comment]
Ok. Well I thought I could do this sometimes later today, if not another day this week. But in the end, I thought I could just get rid of this task while having my coffee. I pushed a few commits and fixed all the mentioned issues.

**Comment by Jehan** [comment]
Note: there are still a few issues because the make distcheck fails. I will take care of these later.

**Comment by Jehan** [comment]
I have mostly fixed everything. There remains only 1 problem but I can't seem to find a solution yet: POTFILES.in only looks into the source directory, not the build directory (apparently the logic is that [translators should be able to check the sources without building](https://wiki.gnome.org/TranslationProject/DevGuidelines/Localize%20using%20gettext%20and%20intltool)).

So this is ok when building in the same directory but is a problem for VPATH builds (hence `make distcheck` as well). The solution would be to extract the strings directly from the json file, but I can't find the right solution for this.

From tomorrow, I will be in the Alps mountain for a few days with probably no internet access so this last fix will have to wait a week or so.

**Comment by achadwick** [comment]
@Jehan 
I pushed a few little tweaks, mostly geared towards making the .so/.dll fully self-contained. I hope I haven't introduced too much nastiness.

`*-gen.h` are generated, certainly, but let's avoid depending on Python during the build for that. They're nigh-static, and don't depend on any other config, so they can reside quite happily in the main repo just as at present on master. However the maintainer will need to regenerate them when `brushsettings.json` changes.

Can we get to a stage where `autoreconf -vfi` is all that's needed? Then the generate script can be stripped down. For some reason, this seems a little nicer to do on Windows/MSYS2.

Debianization repo created: <https://github.com/mypaint/libmypaint.deb>.

Getting a little closer on getting this thing working on Win32/MSYS2, but it's proving painful on that platform due to some nasty-looking redirection or macro expansion errors. I'll keep you informed, but we have to be able to build MSYS2 packages before release.

**Comment by achadwick** [comment]
I'm putting together a native Windows PKGBUILD via MSYS2 ([main repo](https://github.com/Alexpux/MINGW-packages)). [PKGBUILD](https://github.com/achadwick/MINGW-packages/blob/fd3e51bb98ff81eb0672fc4b42c4ea9c06651ad2/mingw-w64-libmypaint-git/PKGBUILD).

It works great for `x86_64`, but `i686` barfs in `build()` with:

```
[...]
libtool: link: gcc -shared  .libs/brushmodes.o .libs/fifo.o .libs/helpers.o .libs/mapping.o .libs/mypaint.o .libs/mypaint-brush.o .libs/mypaint-brush-settings.o .libs/mypaint-fixed-tiled-surface.o .libs/mypaint-rectangle.o .libs/mypaint-surface.o .libs/mypaint-tiled-surface.o .libs/operationqueue.o .libs/rng-double.o .libs/tilemap.o .libs/utils.o   -LC:/msys64/mingw32/lib -ljson-c  -march=i686 -mtune=generic -O2 -ggdb -Og -ggdb -Og   -o .libs/libmypaint-0.dll -Wl,--enable-auto-image-base -Xlinker --out-implib -Xlinker .libs/libmypaint.dll.a
.libs/mypaint-brush-settings.o: In function `mypaint_brush_setting_info_get_name':
C:\msys64\usr\src\MINGW-packages\mingw-w64-libmypaint-git\src\build-i686-w64-mingw32/mypaint-brush-settings.c:49: undefined reference to `libintl_dgettext'
.libs/mypaint-brush-settings.o: In function `mypaint_brush_setting_info_get_tooltip':
C:\msys64\usr\src\MINGW-packages\mingw-w64-libmypaint-git\src\build-i686-w64-mingw32/mypaint-brush-settings.c:55: undefined reference to `libintl_dgettext'
.libs/mypaint-brush-settings.o: In function `mypaint_brush_input_info_get_name':
C:\msys64\usr\src\MINGW-packages\mingw-w64-libmypaint-git\src\build-i686-w64-mingw32/mypaint-brush-settings.c:81: undefined reference to `libintl_dgettext'
.libs/mypaint-brush-settings.o: In function `mypaint_brush_input_info_get_tooltip':
C:\msys64\usr\src\MINGW-packages\mingw-w64-libmypaint-git\src\build-i686-w64-mingw32/mypaint-brush-settings.c:87: undefined reference to `libintl_dgettext'
collect2.exe: error: ld returned 1 exit status
Makefile:690: recipe for target 'libmypaint.la' failed
make[2]: *** [libmypaint.la] Error 1
make[2]: Leaving directory '/usr/src/MINGW-packages/mingw-w64-libmypaint-git/src/build-i686-w64-mingw32'
Makefile:862: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/usr/src/MINGW-packages/mingw-w64-libmypaint-git/src/build-i686-w64-mingw32'
Makefile:599: recipe for target 'all' failed
```

Not sure why that is. libintl seems to be available, configured in etc.

**Comment by Jehan** [comment]
Hi @achadwick ! Back from 4 days in tent under the snow with no internet connection. :-)

Anyway I can't tell you about your issue because I don't build on Windows (work/coding/shell environment on Windows is just too much on a pain to impose this on myself). I can tell you that I successfully cross-compile libmypaint with autotools now, both 32 and 64 bits in under a minute. It just works, out of the box now! :-)

Why you have a problem with a native build, I have no idea. Are you sure that you are not using the same 64-bit libintl for both the x86_64 and i686 builds?

**Comment by Jehan** [comment]
> Can we get to a stage where autoreconf -vfi is all that's needed? Then the generate script can be stripped down. For some reason, this seems a little nicer to do on Windows/MSYS2.

You mean instead of the autogen.sh script? I kind of like this autogen.sh script because then we don't have to remember how to generate the `configure` script, which command line options to use, and also it does more than just this. But of course this is not mandatory. It's not part of the `./configure && make && make install` triptych and is only a conveniency script for maintainers.

You are the boss and can decide to get rid of it. I'll just say I'd prefer to keep it. Of course, maybe it can be simplified and more adapted to libmypaint. I reused most of the code from the autogen of GIMP without thinking too much, and clearly there are stuff which may not be adapted (but I see you removed some of these already).

**Comment by achadwick** [comment]
@Jehan Sounds kinda idyllic. The lack of internet, I mean :smile: 

Nah, you're the expert on this. I'll defer to whatever you think's normal.

I think my only objection to the autogen script now is that it runs configure directly. For PKGBUILD environments like the msys2 one, it's more better to keep prepare() stuff separate from configure() stuff. Can it be twiddled to get maintainers/developers to call configure themselves?

Currently working on disentangling brushsettings.py from generate.py. If I do it right, MyPaint won't need to consult brushsettings.json at runtime; hopefully it can just use libmypaint's exposed symbols and API calls. That means no need for the $PREFIX/share/libmypaint tree.

**Comment by Jehan** [comment]
@achadwick I actually think that the "running configure" part is annoying too and prefer scripts with well separated tasks. I dislike it in GIMP too but just kept it as is when copying this script to libmypaint (I figure that some people like this). I can remove this part if you want, no prob. :-)

**Comment by Jehan** [comment]
@achadwick Done. I removed the call to `configure` from `autogen.sh` and pushed it. They are clearly dedicated scripts now, and `autogen.sh` is only developer side while `configure` is builder side.

**Comment by Jehan** [comment]
@achadwick This night, I made some fix and now the build goes until `make distcheck` with success. Basically, I don't know how much of the autotools convenience you know about, so sorry if I give you an explanation you already know:

- `make dist` creates 2 archives automatically: `libmypaint-1.3.0.tar.bz2` and `libmypaint-1.3.0.tar.xz`, ready for release. This is *the* way you have to make the release tarballs. The choice of compression is done in configure.ac: `AM_INIT_AUTOMAKE([dist-bzip2 dist-xz no-dist-gzip])` tells automake to create a bzip2 and a xz archives on `make dist`, but no gzip. Feel free to choose other compressions.

- `make distcheck` is checking that everything is fine by creating the archives (so basically it implies `make dist`) then uncompressing it in a temporary directory and VPATH-building from the uncompressed files, and not from the repository! This way, it makes sure that nothing has been forgotten in the archives (in particular if you didn't forget any needed file in EXTRA_DIST).

In other words, it means that you should simply run `make distcheck` before any release. If it fails, something has to be fixed. If it succeeds, the following message will be outputted:

> ==================================================
> libmypaint-1.3.0 archives ready for distribution: 
> libmypaint-1.3.0.tar.bz2
> libmypaint-1.3.0.tar.xz
> ==================================================

And you can just upload the generated archives, and done!
Note also that if you have any continuous integration, you should have it run `make distcheck` as well (better discover any issue as soon as possible!).

**Comment by achadwick** [comment]
Woohoo! Thanks for doing this. Seems to be really moving ahead now!

Can I ask you a quick question about versioning? Elsewhere, MyPaint uses the [SemVer](http://semver.org/) way of marking prereleases. It's easy to slot a "-alpha.0" into `LIBMYPAINT_VERSION` via a `libmypaint_prerelease_version` macro, but should I? Is it likely to break generated pkgconfig or gir files?

**Comment by Jehan** [comment]
I have no idea if pkg-config uses semver, but searching the web, some people say that it doesn't support it (because it would actually be older than the semantic versionning specification, though it is strange that they never patched it since that time!).

So I guess that yes, it would break pkg-config (well only for version comparison. Not for simple libmypaint detection). You should probably keep 2 version variables (a semantic one with optional pre-release tags and another one without) and use the version without pre-release into the pc file. Actually the usage of autotools does not change a thing about this issue. What were you doing before?

**Comment by achadwick** [comment]
> What were you doing before?

Haha basically not caring about pkgconfig that's what :sweat_smile:
MyPaint currently builds libmypaint using SCons and a submodule, statically linking libmypaint into its (C++) Python extension. This is due to change very soon, as you'd expect.

We can take the patterns we've used before for monolithic-MyPaint as the Way To Do Things:

* Tarball versions have the trailing prerelease junk for prereleases → `libmypaint-1.3.0-alpha.0.tar.xz`
* The .pc files say just `1.3.0` even for prereleases. Hope that's OK.
* MyPaint uses tagged releases in github for distribution, and so will libmypaint. The tarball will go there, and the associated tag will be named v1.3.0-alpha.0.
* Downstream builds get to tack on extra version details, e.g. the number of revisions post-tag, and structure the version in a way that makes sense for their versioning strategy e.g.:
  * Debian/Ubuntu PPA: 1.3.0~alpha.0 ([allows 1.3.0 to replace it automatically](http://www.fifi.org/doc/debian-policy/policy.html/ch-versions.html))
  * MSYS2/Arch PKGBUILD that pulls from git: 1.3.0alpha.0.13 ([same deal](https://www.archlinux.org/pacman/pacman.8.html#_operations), this time with a post-tag revision count)

That's what I'm hoping to achieve, so I'm really just checking that it's OK for the pkgconfig file's version to be a little bit ahead of the prerelease's "true" version number. At that point, everything should be both API and ABI stable *by definition*, so it shouldn't matter...

(I really want to avoid nasty hacks like using high patch versions (e.g. 1.2.101) for prereleases.)

**Comment by Jehan** [comment]
Note that I haven't tested. Only did a web search. Maybe pkg-config is able to cope with prerelease tags. It is worth a try.

If it really can't, then I'd say that using the stable tag is better than nothing. In any case, anyone who install pre-release versions is supposed to know what one is doing. Basic users should not install pre-releases.

**Comment by achadwick** [comment]
The maintainer's 2010 guide to pkg-config says that it [uses the RPM algorithm](https://people.freedesktop.org/~dbn/pkg-config-guide.html#writing). RPM [gained support for tilde (~) prerelease notation](http://rpm.org/ticket/56) only recently, in 2012. The pkg-config guide also states:

> This [implementation of rpmvercmp] works best with a dotted decimal number such as 1.2.3 since letters can cause unexpected results.

So, I'm not going to rely on any notation for prerelease tagging working in pkg-config files yet :confused:

Hopefully it's OK to just use the forthcoming release's version. Maybe the (human-readable) Description in the file could be annotated specially if it's a prerelease, however.

**Comment by achadwick** [comment]
I have merged [autotools-port](https://github.com/mypaint/libmypaint/tree/autotools-port) into master with 16693fb9c702e79a39708275a44be90501ad1e78, along with a few bugfixes and build tweaks for MyPaint. The branch can be deleted at our leisure.

Thanks for the awesome contribution, @jehan!
